### PR TITLE
Remove dependence on y_loc when not needed and accelerate get head indices 

### PR DESCRIPTION
**Pull request by pzhanggit**

### CODE REVIEW

**Comment by pzhanggit** [comment]
One thing observed in this PR is that something needs to fixed about profiling in md17 and qm9 in CI test. The error message is 
`
_____________________________ pytest_examples[qm9] _____________________________
Traceback (most recent call last):
  File "/home/runner/work/HydraGNN/HydraGNN/tests/test_examples.py", line 26, in pytest_examples
    assert return_code == 0
AssertionError: assert -11 == 0
----------------------------- Captured stderr call -----------------------------
Downloading https://data.pyg.org/datasets/qm9_v3.zip
Extracting dataset/qm9/raw/qm9_v3.zip
Processing...
Using a pre-processed version of the dataset. Please install 'rdkit' to alternatively process the raw data.
Done!
0: Using CPU
0: Using CPU

  0%|          | 0/11 [00:00<?, ?it/s]
 36%|███▋      | 4/11 [00:00<00:00, 32.53it/s]ERROR:2022-05-12 13:46:01 3754:3754 CudaDeviceProperties.cpp:26] cudaGetDeviceCount failed with code 35
____________________________ pytest_examples[md17] _____________________________
Traceback (most recent call last):
  File "/home/runner/work/HydraGNN/HydraGNN/tests/test_examples.py", line 26, in pytest_examples
    assert return_code == 0
AssertionError: assert -11 == 0
----------------------------- Captured stderr call -----------------------------
`

File: hydragnn/train/train_validate_test.py [hunk 1]
```diff
@@ -194,12 +194,47 @@ def get_head_indices(model, data):
...
+        return head_index
+    # dimensions of all heads
+    head_dims = model.module.head_dims
+    head_dimsum = sum(head_dims)
+
+    batch_size = data.batch.max() + 1
+    for ihead in range(model.module.num_heads):
+        head_each = torch.arange(head_dims[ihead])
+        head_ind_temporary = head_each.repeat(batch_size)
+        head_shift_temporray = torch.repeat_interleave(
```
Comment (allaffa) [inline hydragnn/train/train_validate_test.py:220]:
```suggestion
        head_shift_temporary = torch.repeat_interleave(
```

File: hydragnn/train/train_validate_test.py [hunk 2]
```diff
@@ -194,12 +194,47 @@ def get_head_indices(model, data):
...
+            torch.arange(batch_size) * head_dimsum, head_dims[ihead]
+        )
+        head_index[ihead] = head_ind_temporary + head_shift_temporray
```
Comment (allaffa) [inline hydragnn/train/train_validate_test.py:223]:
```suggestion
        head_index[ihead] = head_ind_temporary + head_shift_temporary
```

**Comment by allaffa** [comment]
> One thing observed in this PR is that something needs to fixed about profiling in md17 and qm9 in CI test. @jychoi-hpc @streeve The error message is ` _____________________________ pytest_examples[qm9] _____________________________ Traceback (most recent call last): File "/home/runner/work/HydraGNN/HydraGNN/tests/test_examples.py", line 26, in pytest_examples assert return_code == 0 AssertionError: assert -11 == 0 ----------------------------- Captured stderr call ----------------------------- Downloading https://data.pyg.org/datasets/qm9_v3.zip Extracting dataset/qm9/raw/qm9_v3.zip Processing... Using a pre-processed version of the dataset. Please install 'rdkit' to alternatively process the raw data. Done! 0: Using CPU 0: Using CPU
> 
> 0%| | 0/11 [00:00<?, ?it/s] 36%|███▋ | 4/11 [00:00<00:00, 32.53it/s]ERROR:2022-05-12 13:46:01 3754:3754 CudaDeviceProperties.cpp:26] cudaGetDeviceCount failed with code 35 ____________________________ pytest_examples[md17] _____________________________ Traceback (most recent call last): File "/home/runner/work/HydraGNN/HydraGNN/tests/test_examples.py", line 26, in pytest_examples assert return_code == 0 AssertionError: assert -11 == 0 ----------------------------- Captured stderr call ----------------------------- `

@pzhanggit  
I am trying to gage where we are with this PR. 
I have one question, if the CI test returns this error, why is the PR labeled with a green check as if all the CI tests completed successfully?

**Comment by pzhanggit** [comment]
> 

The error was there when we use the profiling tools on CPU for qm9 and md17. The test is successful because we are not using the profiling tools now.

**Review by jychoi-hpc — APPROVED** [review]
It looks good to me. I tested on my local machine and it works.

**Review by allaffa — APPROVED** [review]

**Comment by allaffa** [comment]
@jychoi-hpc @pzhanggit 
Before merging, do we want to make sure that the changes make the CI workflow behave well also with profiling tools?

**Comment by pzhanggit** [comment]
> @jychoi-hpc @pzhanggit Before merging, do we want to make sure that the changes make the CI workflow behave well also with profiling tools?

Yes, but the error we observed about profiling in qm9 and md17 is not due to this branch. Instead, I think something is not setup properly for cpu mode in profiling part. @jychoi-hpc  Could you confirm this? Thank. I opened a new issue about it, https://github.com/ORNL/HydraGNN/issues/141

**Comment by jychoi-hpc** [comment]
I just tested on my local machine which has only CPU and it went ok. 

I think Sam has put some routine to check CPU/GPU when profiling: https://github.com/pzhanggit/HydraGNN/blob/58a0e7ad9937bca7dc54cb9c1db0795e1cfcbd0d/hydragnn/utils/profile.py#L19

Not sure how the error happened.

**Review by jychoi-hpc — APPROVED** [review]

**Comment by pzhanggit** [comment]
> I just tested on my local machine which has only CPU and it went ok.
> 
> I think Sam has put some routine to check CPU/GPU when profiling: https://github.com/pzhanggit/HydraGNN/blob/58a0e7ad9937bca7dc54cb9c1db0795e1cfcbd0d/hydragnn/utils/profile.py#L19
> 
> Not sure how the error happened.

Interesting. Let me add the profiling back tmr and see if we can reproduce the error.

**Comment by pzhanggit** [comment]
@jychoi-hpc  It passed on my local workstation with cpu only, but failed again here with profiling on in qm9 and md17. Do you know what is going on?

File: hydragnn/train/train_validate_test.py (hunk 1 above)
Comment (pzhanggit) [reply hydragnn/train/train_validate_test.py]: Done

**Comment by allaffa** [comment]
@pzhanggit @jychoi-hpc 
Is this ready to merge?

**Pull request merged**